submariner/IPSec-transport-mode.md

```diff
@@ -0,0 +1,106 @@
+
+# Cable Driver Enhancement - IPSec Transport Mode
+
+## Summary
+
+Submariner today supports three cable drivers: VXLAN, IPSec (tunnel mode) and WireGuard. The
+goal of this proposal is to outline the enhancements needed for the Submariner
+Cable Driver to enable IPSec in transport mode.
+
+For Overlay-over-IPSec in transport mode with further configuration, the benefits fall into the
+following categories:
+
+* **Performance**: Taking advantage of an overlay protocol wrapped in offloaded IPSec in transport mode can
+yield significant performance improvements. The existing IPSEC configuration performance on a pair of modern
+servers ~ 2Gbit maximum. Switching to IP GRE Tunnel protected by an offloaded transport mode IPSEC external
+“wrap” - 25Gbit. A difference of 12.5 times. Or in the case of VXLAN offloaded with an IPSec transport wrap - 9Gbit.
```

Comment:
Yes, @kot-begemot-uk ran these benchmarks and has a report - he will be able to share on return from PTO

Reply:
I ran a set of tests on ixgbe (most popular 10G adapter - > 50% market) and Melanox. You can run them yourself: https://github.com/kot-begemot-uk/tunnel-tests

Tunnel mode offload is totally disabled in the ixgbe upstream driver at present (I can dig out the exact commit, it is somewhere in 5.x and got backported to all -stable series). It can be enabled on TX, but not on RX - I submitted a patch to the kernel, but it got ignored. The maintainer decided that it is safer just to keep it off completely. Even with TX-only enabled you get at most 5G, usually less.

All in all, the performance in tunnel mode on most cards likely to be found in an SME and basic bare metal hosting is guaranteed to be abysmal. 

Compared to that transport mode works both on ixgbe and Melanox for supported adapters. As a result you get anything between 6-9G and line rates where offload works for all encaps levels. Up to 20 times performance difference in some cases.

Reply:
This seems like an advantage for transport mode and I guess we can benefit from HW offload in BM K8s (k8s node == BM server) deployments,

but is HW offload also applicable in public clouds 'virtual machines' and OSP VMs?

IIRC, to consume HW offload capabilities the 'vm' needs to be connected 'directly' (for example: sriov) to the NIC.

Reply:
@kot-begemot-uk thanks for sharing, Can this supporting evidence be added in some form to the EP?
Either as a sub section or if you have some place else then a link to the results and how to reproduce them would be good.